### PR TITLE
[BEAM-14166] Performance improvements for RowWithGetter

### PR DESCRIPTION
**Pull request by mosche**
Push field type based logic in `RowWithGetters` down into `FieldValueGetter`s to remove any branching and allow for better inlining and switch to TreeMap for caching to minimize memory footprint. See benchmark results below. 

To not modify the behavior of `getValues()`, related calls are delegated to the original getter as is.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by mosche** [comment]
R: @TheNeuralBit 
R: @reuvenlax 
R: @kennknowles

**Comment by TheNeuralBit** [comment]
> By default, getValue results in generated bytecode to fetch the value. Is this cache really more efficient than the generated bytecode?

Maybe some microbenchmarks could help here

**Comment by mosche** [comment]
@reuvenlax certainly a valid question and I'm happy to discuss what's worth caching. 

Though, the key point here is, that there's already a cache in place for array, iterable and map types. Processing for these is currently not ideal. For instance, looking at an array field the following steps happen on `getValue`:

1) Invoke getter, the generated byte code wraps the array in a list
2) Then lazily transform the list elements (`Lists.transform`) and cache the result

The problem is really on the 2nd access. Step 1) is repeated, but step 2) uses the cache discarding all the work of step 1).

This mostly addresses some of the inconsistencies using the existing cache and reduces the overhead of the cache by choosing better fitted data structures to store cached values.

**Comment by mosche** [comment]
Considering that elements of `ARRAY` types and `ITERABLE` types are [lazily transformed](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/values/RowWithGetters.java#L81-L97), the value of caching is rather small (questionable?) as it happens on demand.

For `MAP` types it's a different story as the entire map gets [rewritten on transform](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/values/RowWithGetters.java#L99-L106).

And if a cache is used, then any composite `ROW` type should be cached. Otherwise, if re-created every time, the cache is obviously pointless.

**Comment by mosche** [comment]
I started doing some jmh benchmarks for `RowWithGetter`. I'll follow up with next week on this...

**Comment by reuvenlax** [comment]
Benchmarks of this code are tricky to do, since the cost of codegen tends to dominate microbenchmarks.

**Comment by mosche** [comment]
I've pushed my [benchmark code ](https://github.com/apache/beam/pull/17203) for reference, let me know if you have any suggestions. To prevent the issue you've mentioned @reuvenlax, I'm setting up a relatively large number of rows as JMH state before the actual benchmark invocation.

To establish a baseline, I'm looking at `master` first. [Here](https://jmh.morethan.io/?sources=https://raw.githubusercontent.com/apache/beam/3b5e1d607d30397abb27a119fc30c5d05a7b2d40/sdks/java/harness/jmh/RowWithGetters-master-cache-disabled.json,https://raw.githubusercontent.com/apache/beam/3b5e1d607d30397abb27a119fc30c5d05a7b2d40/sdks/java/harness/jmh/RowWithGetters-master-cache-enabled-lazy-init.json) are some initial results with some minimal changes to `RowWithGetters` to make these benchmarks meaningful:
- Disable caching for the first run.
- Change initialisation of the cache data structure to lazy init so associated costs are considered in the benchmark.

These numbers are certainly not very much in favour of the status quo.
I'm still iterating on improvements, but from what I've seen so far far there's lots that can be done.

**Comment by mosche** [comment]
I've drilled down into this a bit and I think I've got some interesting finding's to share @reuvenlax & @TheNeuralBit.

Investigating a few approaches, I would to suggest to push field type based logic in `RowWithGetters` down into `FieldValueGetter`s to remove any branching and allow for better inlining, see [code & benchmark](https://github.com/apache/beam/compare/master...mosche:RowWithGetters-JMH-Getters-LogicPushDown).

I also looked a bit into costs of caching, the picture isn't as clear there. The costs of initializing any data structure facilitating a cache is certainly high compared to the costs of calling getters. One finding though was that `TreeMap` didn't perform any worse than `HashMap`. Given the much lower memory footprint that might be a good pick then. Also, using [materialized Pojo lists](https://github.com/mosche/beam/blob/RowWithGetters-JMH-Getters-LogicPushDown/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/GetterBasedSchemaProvider.java#L178-L180) helped to improve the performance gain from caching (compared to lazy transforms using `Lists.transform`).

On the other hand, I'm not sure what the original motivation for adding a field value cache in `RowWithGetters` was. Is it just about performance?

Some visualizations for a few selected runs:
- [master vs suggested approach (no caching)](https://jmh.morethan.io/?sources=https://raw.githubusercontent.com/mosche/beam/RowWithGetters-JMH-Getters-LogicPushDown/sdks/java/harness/jmh/jmh-master-cache-disabled.json,https://raw.githubusercontent.com/mosche/beam/RowWithGetters-JMH-Getters-LogicPushDown/sdks/java/harness/jmh/jmh-pr-getters-cache-disabled.json)
- [master vs suggested approach (with caching)](https://jmh.morethan.io/?sources=https://raw.githubusercontent.com/mosche/beam/RowWithGetters-JMH-Getters-LogicPushDown/sdks/java/harness/jmh/jmh-master-cache-HashMaps.json,https://raw.githubusercontent.com/mosche/beam/RowWithGetters-JMH-Getters-LogicPushDown/sdks/java/harness/jmh/jmh-pr-getters-cache-TreeMap-lazy-lists.json)
- [costs of caching (master)](https://jmh.morethan.io/?sources=https://raw.githubusercontent.com/mosche/beam/RowWithGetters-JMH-Getters-LogicPushDown/sdks/java/harness/jmh/jmh-master-cache-disabled.json,https://raw.githubusercontent.com/mosche/beam/RowWithGetters-JMH-Getters-LogicPushDown/sdks/java/harness/jmh/jmh-master-cache-HashMaps.json)
- [costs of caching (suggested approach, treemap)](https://jmh.morethan.io/?sources=https://raw.githubusercontent.com/mosche/beam/RowWithGetters-JMH-Getters-LogicPushDown/sdks/java/harness/jmh/jmh-pr-getters-cache-disabled.json,https://raw.githubusercontent.com/mosche/beam/RowWithGetters-JMH-Getters-LogicPushDown/sdks/java/harness/jmh/jmh-pr-getters-cache-TreeMap-lazy-lists.json)
- [costs of caching (suggested approach, hashmap)](https://jmh.morethan.io/?sources=https://raw.githubusercontent.com/mosche/beam/RowWithGetters-JMH-Getters-LogicPushDown/sdks/java/harness/jmh/jmh-pr-getters-cache-disabled.json,https://raw.githubusercontent.com/mosche/beam/RowWithGetters-JMH-Getters-LogicPushDown/sdks/java/harness/jmh/jmh-pr-getters-cache-HashMap-lazy-lists.json)
- [hashmap vs treemap if using materialized rather than lazy Pojo collections](https://jmh.morethan.io/?sources=https://raw.githubusercontent.com/mosche/beam/RowWithGetters-JMH-Getters-LogicPushDown/sdks/java/harness/jmh/jmh-pr-getters-cache-HashMap-persisted-pojolists.json,https://raw.githubusercontent.com/mosche/beam/RowWithGetters-JMH-Getters-LogicPushDown/sdks/java/harness/jmh/jmh-pr-getters-cache-TreeMap-persisted-pojolists.json)

[Screenshot 2022-04-01 at 14 49 46](https://jmh.morethan.io/?sources=https://raw.githubusercontent.com/mosche/beam/RowWithGetters-JMH-Getters-LogicPushDown/sdks/java/harness/jmh/jmh-master-cache-HashMaps.json,https://raw.githubusercontent.com/mosche/beam/RowWithGetters-JMH-Getters-LogicPushDown/sdks/java/harness/jmh/jmh-pr-getters-cache-TreeMap-lazy-lists.json)

**Comment by mosche** [comment]
Run Java PreCommit

**Comment by mosche** [comment]
Run Java PreCommit

**Comment by mosche** [comment]
@TheNeuralBit @reuvenlax Please let me know how / if you wanna proceed here, so I can wrap up my initial PR https://github.com/apache/beam/pull/16947 accordingly. In any case, it was a very interesting dive into Beam Schemas. Thanks for all the pointers.

**Comment by mosche** [comment]
@TheNeuralBit  @reuvenlax ping

**Comment by aromanenko-dev** [comment]
@TheNeuralBit @reuvenlax 
Kind ping on this

**Comment by TheNeuralBit** [comment]
Run SQL PostCommit

**Comment by TheNeuralBit** [comment]
Run Java PreCommit

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldValueGetter.java [hunk 1]
```diff
@@ -33,5 +33,9 @@
   @Nullable
   ValueT get(ObjectT object);
 
+  default @Nullable Object getRaw(ObjectT object) {
```
Comment (TheNeuralBit) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldValueGetter.java:36]: nit: maybe `getUntyped` is a more appropriate name here?

Also could you clarify why we need this?

**Review by TheNeuralBit — APPROVED** [review]
Thanks @mosche for the extensive benchmarking, and apologies for the atrocious review latency!

I'm ok with this if it passes current tests, given the extensive benchmarking. A couple of things:
- I'd still like to get some final feedback on the approach from @reuvenlax, but I acn go ahead and merge next Monday if that doesn't happen this week.
- It could be interesting to run your benchmarks continuously to 1) monitor for regressions, and 2) give us a convenient way to test other potential improvements. Maybe file a jira for that if you don't have time now?

**Comment by reuvenlax** [comment]
Let me take another look. IIRC I had some concerns, but it's possible that
the extensive benchmarking addressed those.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldValueGetter.java (hunk 1 above)
Comment (mosche) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldValueGetter.java:36]: Thanks so much for having a look @TheNeuralBit 🙏 

`getRaw()` was based on a [conversation](https://lists.apache.org/thread/gjq453fm32s76zlvjs4kb5g3rgxnh7gs) with @reuvenlax.

> getValues() is maybe poorly named - might be better called getRawValues. What you're  looking for is  probably the getBaseValues() method.
> getValues is mostly used in code that knows exactly what it's doing for  optimization purposes. It goes along with the attachValues method, which is similarly tricky to use. It's there to enable  0-copy  code, but not necessarily intended for general consumption. 

`RowWithGetters.getValues()`  returns the "raw" unmodified result of the getters:
```java
public List<Object> getValues() {
  return getters.stream().map(g -> g.get(getterTarget)).collect(Collectors.toList());
}
```

As I am pushing down the transformation of the getter result into the getter itself, I needed a way to bypass that in order to maintain the current semantics of `getValues()`. Let me know if the name makes sense given that context.

Comment (TheNeuralBit) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldValueGetter.java:36]: Yeah that makes sense, thanks. Could you add some of this context in a comment there?

**Comment by mosche** [comment]
@reuvenlax Kind ping, did you already get the chance to have another look at this? Just to make it clear, this is not about caching ... that was just a minor almost irrelevant part of this

**Comment by aaltay** [comment]
@reuvenlax - Would you be able to take another look?

**Comment by TheNeuralBit** [comment]
I think this is fine to go ahead and merge. From what I can tell Reuven's original concern was that this would start caching more types than were already being cached in `RowWithGetters`, but this is just moving around the caching logic.

**Comment by aromanenko-dev** [comment]
@TheNeuralBit Thanks for moving forward and finally merge it!

**Comment by mosche** [comment]
@TheNeuralBit Thanks a lot 🙏  I still owe you the comment on `FieldValueGetter.getRaw`, i'll follow up shortly!

**Comment by mosche** [comment]
Resolves #21634

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldValueGetter.java (hunk 1 above)
Comment (mosche) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldValueGetter.java:36]: @TheNeuralBit I've opened a new PR to add the missing comment, sorry for the delay.
https://github.com/apache/beam/pull/21982

**Comment by reuvenlax** [comment]
Will look at it today or tomorrow - currently on a flight.

**Pull request merged**